elasticgraph-graphql/lib/elastic_graph/graphql/filtering/filter_interpreter.rb

```diff
@@ -102,13 +197,20 @@ def filters_on_sub_fields?(expression)
           end
         end
 
+        def process_empty_or_nil_expression(bool_node, field_or_op)
+          if field_or_op == schema_names.not
```

Comment:
It's a code smell that the handling for `not` is spread out--it's handled primarily by `process_not_expression` but also by this method.  we want each different type of filtering node to be handled in one place and not have its responsibility spread out.

I understand why you did this, though: the current `FilterNodeInterpreter` implementation does not allow the `:empty` case to _just_ handle the empty case and allow the `:not` case to just handle the not case, because it returns `:empty` for both `field: nil` and `not: nil`.  The current implementation looks like this:

```ruby
def identify_node_type(field_or_op, sub_expression)
  return :empty if sub_expression.nil? || sub_expression == {}
  return :not if field_or_op == schema_names.not
  return :list_any_filter if field_or_op == schema_names.any_satisfy
  return :all_of if field_or_op == schema_names.all_of
  return :any_of if field_or_op == schema_names.any_of
  return :operator if filter_operators.key?(field_or_op)
  return :list_count if field_or_op == LIST_COUNTS_FIELD
  return :sub_field if sub_expression.is_a?(::Hash)
  :unknown
end
```

Notice that it detects `:empty` based on `sub_expression` before it has the chance to detect `:not` based on `field_or_op`.  That robs `process_not_expression` from the chance to negate the inner empty expression.

I'm thinking that we should always do the detection from the "outside in". That is, `field_or_op` is always on the outside and we should first detect a node type based on that. Only once we've exhausted what we can know based on `field_or_op` should we detect the type based on the `sub_expression`.  I think the implementation should instead be something like this:

```ruby
def identify_node_type(field_or_op, sub_expression)
  identify_by_field_or_op(field_or_op) || identify_by_sub_expr(sub_expression) || :unknown
end

private

def identify_by_field_or_op(field_or_op)
  return :not if field_or_op == schema_names.not
  return :list_any_filter if field_or_op == schema_names.any_satisfy
  return :all_of if field_or_op == schema_names.all_of
  return :any_of if field_or_op == schema_names.any_of
  return :operator if filter_operators.key?(field_or_op)
  return :list_count if field_or_op == LIST_COUNTS_FIELD
end

def identify_by_sub_expr(sub_expression)
  return :empty if sub_expression.nil? || sub_expression == {}
  return :sub_field if sub_expression.is_a?(::Hash)
end
```

Now, that change on its own probably breaks some stuff (the various `process_*` methods haven't had to handle empty or `nil` sub-expressions before because the `:empty` branch took precedence) but it could be worthy prepatory refactoring in its own right to first fix `identify_node_type` in its own PR, and then leverage it here so that this can just return `match_all` and `process_not_expression` can simply negate it.

Thoughts?

Reply:
This makes sense. Originally, I thought of changing the order of when :not and :empty is returned, so this can be handled in the corresponding method. However, I think this is a better approach, where each method is handling this individually.